test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs

```diff
@@ -72,6 +74,165 @@ public class IISMiddlewareTests
+                {
+                    var appLifetime = app.ApplicationServices.GetRequiredService<IApplicationLifetime>();
+                    appLifetime.ApplicationStopping.Register(() => applicationStoppingFired.Set());
+
+                    app.Run(context =>
+                    {
+                        requestExecuted.Set();
+                        return Task.FromResult(0);
+                    });
+                });
+            var server = new TestServer(builder);
+
+            var request = new HttpRequestMessage(HttpMethod.Post, requestPath);
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-TOKEN", "TestToken");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-EVENT", shutdownEvent);
+            var response = await server.CreateClient().SendAsync(request);
+
+            Assert.False(requestExecuted.WaitOne(0));
+            Assert.True(applicationStoppingFired.WaitOne(TimeSpan.FromSeconds(5)));
+            Assert.Equal(HttpStatusCode.Accepted, response.StatusCode);
+        }
+
+        public static TheoryData<HttpMethod> InvalidShutdownMethods
+        {
+            get
+            {
+                return new TheoryData<HttpMethod>
+                {
+                    HttpMethod.Put,
+                    HttpMethod.Trace,
+                    HttpMethod.Head,
+                    HttpMethod.Get,
+                    HttpMethod.Delete,
+                    HttpMethod.Options
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(InvalidShutdownMethods))]
+        public async Task MiddlewareIgnoresShutdownGivenWrongMethod(HttpMethod method)
+        {
+            var requestExecuted = new ManualResetEvent(false);
+            var applicationStoppingFired = new ManualResetEvent(false);
+            var builder = new WebHostBuilder()
+                .UseSetting("TOKEN", "TestToken")
+                .UseSetting("PORT", "12345")
+                .UseSetting("APPL_PATH", "/")
+                .UseIISIntegration()
+                .Configure(app =>
+                {
+                    var appLifetime = app.ApplicationServices.GetRequiredService<IApplicationLifetime>();
+                    appLifetime.ApplicationStopping.Register(() => applicationStoppingFired.Set());
+
+                    app.Run(context =>
+                    {
+                        requestExecuted.Set();
+                        return Task.FromResult(0);
+                    });
+                });
+            var server = new TestServer(builder);
+
+            var request = new HttpRequestMessage(method, "/iisintegration");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-TOKEN", "TestToken");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-EVENT", "shutdown");
+            var response = await server.CreateClient().SendAsync(request);
+
+            Assert.True(requestExecuted.WaitOne(0));
```

Comment:
0 is weird for these, it assumes the request was executed synchronously. Use 5 seconds.